internal/elasticsearch/logstash/pipeline.go

```diff
@@ -0,0 +1,259 @@
+
+	logstashPipeline.PipelineSettings = map[string]interface{}{}
+	if settings := expandIndividuallyDefinedPipelineSettings(ctx, d, allSettingsKeys); len(settings) > 0 {
+		logstashPipeline.PipelineSettings = settings
+	}
+
+	logstashPipeline.Username = d.Get("username").(string)
+
+	if diags := client.PutLogstashPipeline(ctx, &logstashPipeline); diags.HasError() {
+		return diags
+	}
+
+	d.SetId(id.String())
+	return resourceLogstashPipelineRead(ctx, d, meta)
+}
+
+func resourceLogstashPipelineRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, err := clients.NewApiClient(d, meta)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	resourceID, diags := clients.ResourceIDFromStr(d.Id())
+	if diags.HasError() {
+		return diags
+	}
+
+	logstashPipeline, diags := client.GetLogstashPipeline(ctx, resourceID)
+	if logstashPipeline == nil && diags == nil {
+		d.SetId("")
+		return diags
+	}
+	if diags.HasError() {
+		return diags
+	}
+
+	if err := d.Set("pipeline_id", logstashPipeline.PipelineID); err != nil {
+		return diag.FromErr(err)
+	}
+	if err := d.Set("description", logstashPipeline.Description); err != nil {
+		return diag.FromErr(err)
+	}
+	if err := d.Set("last_modified", logstashPipeline.LastModified); err != nil {
+		return diag.FromErr(err)
+	}
+	if err := d.Set("pipeline", logstashPipeline.Pipeline); err != nil {
+		return diag.FromErr(err)
+	}
+	if err := d.Set("pipeline_metadata", logstashPipeline.PipelineMetadata); err != nil {
+		return diag.FromErr(err)
+	}
+	for key, typ := range allSettingsKeys {
+		var value interface{}
+		if v, ok := logstashPipeline.PipelineSettings[key]; ok {
+			value = v
+		} else {
+			tflog.Warn(ctx, fmt.Sprintf("setting '%s' is not currently managed by terraform provider and has been ignored", key))
+			continue
+		}
+		switch typ {
+		case schema.TypeInt:
+			value = int(math.Round(value.(float64)))
+		}
+		if err := d.Set(convertSettingsKeyToTFFieldKey(key), value); err != nil {
+			return diag.FromErr(err)
+		}
+	}
+	if err := d.Set("username", logstashPipeline.Username); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return nil
+}
+
+func resourceLogstashPipelineDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, err := clients.NewApiClient(d, meta)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	resourceID, diags := clients.ResourceIDFromStr(d.Id())
+	if diags.HasError() {
+		return diags
+	}
+
+	if diags := client.DeleteLogstashPipeline(ctx, resourceID); diags.HasError() {
+		return diags
+	}
+	return nil
+}
+
+func expandIndividuallyDefinedPipelineSettings(ctx context.Context, d *schema.ResourceData, settingsKeys map[string]schema.ValueType) map[string]interface{} {
+	settings := make(map[string]interface{})
+	for key := range settingsKeys {
+		tfFieldKey := convertSettingsKeyToTFFieldKey(key)
+		settings[key] = d.Get(tfFieldKey)
```

Comment:
With the `pipeline.workers` comment above this should likely be something like
```suggestion
		if setting, ok := d.GetOk(tfFieldKey); ok { 
			settings[key] = setting
		}
```

I wonder if we should instead extract this method into utils and re-use the index implementation instead.

Reply:
Yeah I have no clue why I copied and simplified this, have extracted to utils now though as suggested 👍